test/evp_pkey_provided_test.c

```diff
@@ -550,6 +554,37 @@ static int test_fromdata_ec(void)
         || !TEST_int_eq(EVP_PKEY_size(pk), 2 + 35 * 2))
         goto err;
 
+    char pub[sizeof(ec_pub_keydata)];
+    char priv[sizeof(ec_priv_keydata)];
+    char curve_name[80];
+    OSSL_PARAM todata_params[4];
+    todata_params[0] =
+        OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_EC_NAME, curve_name,
+                                         sizeof(curve_name));
+    todata_params[1] =
+        OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_PUB_KEY, pub,
+                                         sizeof(pub));
+    todata_params[2] =
+        OSSL_PARAM_construct_BN(OSSL_PKEY_PARAM_PRIV_KEY, (unsigned char *)priv,
+                                sizeof(priv));
+    todata_params[3] = OSSL_PARAM_construct_end();
+
+    if (!TEST_true(EVP_PKEY_todata(pk, todata_params)))
+        goto err;
+
+    if (!TEST_str_eq(curve_name, curve))
+        goto err;
+
+    /* The public key is returned is in COMPRESSED FORMAT */
+    if (!TEST_true(pub[0] == (POINT_CONVERSION_COMPRESSED + 1))
+        || !TEST_mem_eq(pub + 1, todata_params[1].data_size - 1,
+                        ec_pub_keydata + 1, compressed_sz - 1))
+        goto err;
+
+    /*
+     * TODO(3.0): How do we compare bignums here?
+     * The value returned in priv is in BE order.
```

Comment:
Er, what?  The value of priv *should* come in whatever byte order a native integer uses.

Reply:
Could we have a seperate type for OSSL_PARAM_TYPE_BIGNUM that doesnt need to do native order....  Then we dont need to do this reversing bytes stuff that is confusing,
(I keep on saying its confusing because IT IS CONFUSING :)).
i.e- fromdata takes a non native unsigned char[], todata currently returns a unsigned char[] with reversed byte order (THIS IS NOT GOOD).

Reply:
I don't think that `BIGNUM` is good at all.  However, since the data type is supposed to describe memory organization, I'd suggest something like `OSSL_PARAM_TYPE_INTEGER_LE`, `OSSL_PARAM_TYPE_UNSIGNED_INTEGER_LE`, `OSSL_PARAM_TYPE_INTEGER_BE` `OSSL_PARAM_TYPE_UNSIGNED_INTEGER_BE`, and either end that cares must deal with the difference.

The reason I don't want to treat BIGNUM as an essentially different type than any other number can be illustrated with RSA's *e*.  That's usually a small number, and it *should be perfectly legitimate for one end to treat it as an `unsigned int` and the other to treat it as a `BIGNUM` without having to jump through extra hoops*.

Reply:
I dont agree with this logic.. 
I think this is dealing with an edge case that is not needed.
If we want to put a constant into a bignum anywhere else in the code we convert it to a bignum.
For the todata operation - creating a unsigned char array shouldnt be a big deal.

It essentially boils down to this for me...
If I feed an array of bytes into a system, I expect it to be able to be extracted as an array of bytes in exactly the same order (not reversed due to some internal implementation details).

There is no way of knowing that a number is an actual bignum currently (only the end point API's convert to and from a  bignum).  While this isnt a problem at the moment we might find a case where it becomes a issue and then we cant fix it because we have shared the UNSIGNED_INT type.  I have no problem if it asserts if a int is assigned to a bignum type.

Reply:
> I'd suggest something like OSSL_PARAM_TYPE_INTEGER_LE, OSSL_PARAM_TYPE_UNSIGNED_INTEGER_LE, 

You are talking about an internal representation which seems irrelevant.
What about OSSL_PARAM_TYPE_BIG_UNSIGNED_INTEGER?

Reply:
Why is BIG so important? Doesn't the data size tell you that it's big relative to whatever types you are using?

The point here is that you cannot know what the other end is doing, what kind of variables they're using etc. You *happen to know* when you're authoring both ends of the transferred data, but that's not always going to be the case. This is also why OSSL_PARAM is a public structure.
The data type is *exactly* about representation, although I call it memory organisation, there's nothing "internal" about that. This is by design.

Reply:
See the bit above that says 'It essentially boils down to this for me...'
That is my issue - the only way to solve that it to know its a bignum. I cant tell otherwise.
The internal representation is affecting the public interface - that is not good.

Reply:
It's a number, represented as an array of bytes in native byte order. This goes for *any* number (integer, to be precise), regardless of size and regardless of the language type you're using for it.

Reply:
Its a array that gets reversed when it comes back out.

Reply:
The point here is that it seems you'd rather have OCTET_STRING to transfer these bits. And that's fair, I don't always see a point of insisting that the key data be seen as numbers, even though they are *internally* treated as such. The EC case is a bit absurd in that regard, with one half of the key being an octet string and the other being a number..

Reply:
exactly - but then it still needs the type so it knows its a bignum right?

Reply:
[sigh]
@slontis, you want something that's not represented in native byte order, but rather specifically in big endian order. I proposed data types for that. How about we go from there?

Reply:
It being used as a BIGNUM on either end is internal information.